cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/LongRunningOperationRequest.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.operation;
+
+import io.cdap.cdap.internal.app.sourcecontrol.PullOperationInput;
+import io.cdap.cdap.proto.id.OperationRunId;
+import io.cdap.cdap.proto.operation.OperationType;
+import javax.annotation.Nullable;
+
+/**
+ * Encapsulates a operation run request.
+ */
+public class LongRunningOperationRequest {
+
+  // The id for the current run of the operation
+  private final OperationRunId runId;
+  // The type of the operation. Used to load the right operation class
+  private final OperationType type;
+
+  // The possible request types. Only one of them can be not null
+  // The operation will read the necessary request type.
+  // Similar to union type / protobuff oneof
+  // TODO(samik) revisit this implementation after 6.10 release
+  @Nullable
+  private final PullOperationInput pullOpInput;
+
+  /**
+   * Default constructor.
+   */
+  private LongRunningOperationRequest(OperationRunId runid, OperationType operationType,
+      @Nullable PullOperationInput pullOpInput) {
+    this.runId = runid;
+    this.type = operationType;
+    this.pullOpInput = pullOpInput;
+  }
+
+  public OperationRunId getRunId() {
+    return runId;
+  }
+
+  public PullOperationInput getPullOpInput() {
+    return pullOpInput;
+  }
+
+  public OperationType getType() {
+    return type;
+  }
+
+  static Builder builder() {
+    return new Builder();
+  }
+
+  private static final class Builder {
+
+    private OperationRunId runId;
+    private OperationType type;
+    @Nullable
+    private PullOperationInput pullOpInput;
+
+    Builder() {
+      this.runId = null;
+      this.type = null;
+      this.pullOpInput = null;
+    }
+
+    public void setRunId(OperationRunId id) {
+      this.runId = id;
+    }
+
+    public void setType(OperationType type) {
```

Comment:
should combine this with the setRequest() method to ensure the type and request always match.